<a href="https://www.kaggle.com/code/sandhyakrishnan02/multiple-regression-dummy-variable-trap-handled?scriptVersionId=100694526" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Dummy variable trap is handled before building linear multiple regression model.

[Introduction To Linear Assumption](#1)

[Different Linear Assumptions](#2)

[Data Set Analysis](#3)

[Dummy Variable Trap](#4)

[Mathematical Analysis of Dummy Variable Trap](#5)

[Regression Model with Python to handle the dummy variable trap](#6)

[Visualisation](#7)

[Conclusion](#8)

[Reference](#9)



<a id="1"></a> <br>
# 1. Introduction To Linear Assumption

Linear regression is used when the dataset has a linear correlation. Before building a linear regression model assumptions are to be validated. If the assumptions are violated, different methodologies must be used.Simple linear regression has one independent variable (predictor) and a dependent variable(response) and multiple linear regression has more than one predictor to predict response.

The simple linear regression equation is represented as

Y = β0 + β1 (X1 )+ε

Multiple linear regression equation is represented as

Y = β0 + β1 (X1 )+β2 (X 2)+β3(X3)+β4(X4)+ε


<a id="2"></a> <br>
# 2. Different Linear Assumptions

Assumptions of Linear Regression Analysis are :

* Linearity : 
    For linear regression analysis, there must be a linear relationship between predictor(s) and response. Linear relations can be visualized by plotting data points using a scatter plot.
    
* No Heteroskedasticity :
    For a linear relationship if the variance is not a constant and it increases as the predictor (X) increases, then it is called heteroskedasticity. In such cases, the standard error in the output cannot be relied on but still, coefficients will be unbiased.
    
* No omitted variable bias :
    The predictor(s) shouldn’t be correlated with the error term. It can be represented below equation.
    
    𝜎𝑋𝜀= 0: ∀𝑥,𝜀

    If predictor(s) are correlated it is called omitted variable bias. It happens when a predictor is excluded and it gets reflected in the error term which leads to biased and counterintuitive estimates. It can be detected by checking the correlation between predictors.
    
* Normality of error :
    We assume error ε, is normally distributed, i.e, the mean is equal to zero ( the sum of all errors will be equal to zero (0) or almost 0) and variance 𝜎2 of error terms are equal.
    
    ε~ N(0,𝜎2)
    
    Normality is violated when the variance of the errors is not consistent across observations.
    
* No autocorrelation :
    Error term values should not have any identifiable relationship.
    
    𝜎𝜀𝑖𝜀𝑗= 0: ∀𝑖≠𝑗
    
    If there exists any relationship between values of error term then autocorrelation comes into the picture. Autocorrelation is also known as serial correlation which affects the standard errors without affecting coefficients that will be unbiased. 

* No multicollinearity :
   Multicollinearity occurs when predictors are themselves correlated which leads to Dummy Variable Trap. To find if multicollinearity exists between predictors correlation between all pairs of predictors is to be checked. In case the correlation coefficient is very high it clearly indicates the predictor is explained by another predictor.

<a id="3"></a> <br>
# 3. Data Set Analysis 

For this dataset(50_Startups.csv), the first 4 features are used to predict Profit. The First 3 Columns contain numerical data and the fourth column D contains categorical data.

Column D contain 3 US states
1. California
1. Florida
1. New York

<a id="4"></a> <br>
# 4. Dummy Variable Trap

Dummy Variable Trap occurs when predictors are multicollinear, that is one predictor can predict another.

For this dataset, fourth column D contains categorical data. Before building a model categorical data is to be encoded.For encoding column D categorical data, lets us take 3 Dummy Variable 
1. D1 for State_California
1. D2 for State_Florida 
1. D3 for State_New York

Let the Matrix representation of State_California be [100] and State_Florida be [010] and State_New York be [001].

D1, D2 and D3 are highly correlated or multicollinear because if D1 and D2 are zero it is obvious D3 will be 1. ie., if the data does not belong to State_California or State_Florida then it is sure, data belongs to State_NewYork. This means here two dummy variables can predict the third one.

Or in a broader sense, it can be said, if one dummy variable is dropped no relevant information will be lost as other dummy variables can explain dropped feature fully. So if there are n dummy variables, n-1 dummy variables will be able to predict the value of the n-th dummy variable. So in order to avoid the dummy variable trap, one dummy variable has to be dropped.



<a id="5"></a> <br>
# 5. Mathematical Analysis of Dummy Variable Trap

Multiple linear regression equation is represented as

Y = β0 + β1 (X1 )+β2 (X 2)+β3(X3)+β4(X4)+ε

After assigning dummy variable, regression model can be represented as

Y = β0 + β1 (X1 )+β2 (X 2)+β3(X3)+β4D1+β5D2+β6D3+ε — equation 1


we have D1+D2+D3 = 1
or D3 = 1- (D1+D2) — equation 2 which indicates D3 is perfectly explained by the other two dummy variables D1 and D2


If we use equation 2 in 1, we get

Y = β0 + β1 (X1 )+β2 (X 2)+β3(X3)+β4D1+β5D2+β6(1–(D1+D2))+ε

Y = β0 + β1 (X1 )+β2 (X 2)+β3(X3)+β4D1+β5D2+β6-β6D1-β6D2+ε

Y =( β0+ β6)+ β1 (X1 )+β2 (X 2)+β3(X3)+(β4-β6)D1+(β5 -β6)D2+ε

Y = β0*+ β1 (X1 )+β2 (X 2)+β3(X3)+β4*D1+β5*D2 +ε — equation 3


From the equation 3, it is clear that coefficient β0, β4 and β5 is impacted by D3 coefficient β6. So one dummy variable should be dropped to avoid trapped by dummy variable.

<a id="6"></a> <br>
# 6. Regression Model with Python to handle the dummy variable trap

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/startup-logistic-regression/50_Startups.csv


In [2]:
#Importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
%matplotlib inline

import plotly 
import plotly.express as px
import plotly.figure_factory as ff

In [3]:
#Importing dataset
dataset = pd.read_csv('../input/startup-logistic-regression/50_Startups.csv')

In [4]:
#To see first 5 rows of the dataset
dataset.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [5]:
# To see information of dataset
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     object 
 4   Profit           50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


There there is no missing data as all 5 columns are having 50 non-null values. So we are good to go.

In [6]:
#Statistical analysis
dataset.describe()

R&D Spend  Administration  Marketing Spend         Profit
count      50.000000       50.000000        50.000000      50.000000
mean    73721.615600   121344.639600    211025.097800  112012.639200
std     45902.256482    28017.802755    122290.310726   40306.180338
min         0.000000    51283.140000         0.000000   14681.400000
25%     39936.370000   103730.875000    129300.132500   90138.902500
50%     73051.080000   122699.795000    212716.240000  107978.190000
75%    101602.800000   144842.180000    299469.085000  139765.977500
max    165349.200000   182645.560000    471784.100000  192261.830000

In [7]:
dataset.corr()

R&D Spend  Administration  Marketing Spend    Profit
R&D Spend         1.000000        0.241955         0.724248  0.972900
Administration    0.241955        1.000000        -0.032154  0.200717
Marketing Spend   0.724248       -0.032154         1.000000  0.747766
Profit            0.972900        0.200717         0.747766  1.000000

<a id="7"></a> <br>
# 7.Visualisation

## Scatterplotmatrix

In [8]:
fig = ff.create_scatterplotmatrix(dataset, diag='histogram', index='Profit',
                                  height=800, width=800)
fig.show()

## Heat Map

In [9]:
fig = px.imshow(dataset.corr())
fig.show()

From dataset.corr() it is clear Profit is having a high correlation with R & D spend, a good correlation with Marketing Spend and a low correlation with Administration.

## Pie chart

In [10]:
import plotly.graph_objs as go
from plotly.offline import iplot

In [11]:
pie = go.Figure(data=[go.Pie(labels = dataset['State'].value_counts().keys(),values = dataset['State'].value_counts().values)])

iplot(pie)

## Box plot for numerical column

In [12]:
for i, column in enumerate(dataset.columns):
    if i !=3:
        box = go.Figure(go.Box(y = dataset[column], name = column))
        iplot(box)

In [13]:
#Assinging values in X & Y
X = dataset[['R&D Spend', 'Administration', 'Marketing Spend','State']]
y = dataset['Profit']

In [14]:
X

R&D Spend  Administration  Marketing Spend       State
0   165349.20       136897.80        471784.10    New York
1   162597.70       151377.59        443898.53  California
2   153441.51       101145.55        407934.54     Florida
3   144372.41       118671.85        383199.62    New York
4   142107.34        91391.77        366168.42     Florida
5   131876.90        99814.71        362861.36    New York
6   134615.46       147198.87        127716.82  California
7   130298.13       145530.06        323876.68     Florida
8   120542.52       148718.95        311613.29    New York
9   123334.88       108679.17        304981.62  California
10  101913.08       110594.11        229160.95     Florida
11  100671.96        91790.61        249744.55  California
12   93863.75       127320.38        249839.44     Florida
13   91992.39       135495.07        252664.93  California
14  119943.24       156547.42        256512.92     Florida
15  114523.61       122616.84        261776.23    New York
16   78013.11       121597.55        264346.06  California
17   94657.16       145077.58        282574.31    New York
18   91749.16       114175.79        294919.57     Florida
19   86419.70       153514.11             0.00    New York
20   76253.86       113867.30        298664.47  California
21   78389.47       153773.43        299737.29    New York
22   73994.56       122782.75        303319.26     Florida
23   67532.53       105751.03        304768.73     Florida
24   77044.01        99281.34        140574.81    New York
25   64664.71       139553.16        137962.62  California
26   75328.87       144135.98        134050.07     Florida
27   72107.60       127864.55        353183.81    New York
28   66051.52       182645.56        118148.20     Florida
29   65605.48       153032.06        107138.38    New York
30   61994.48       115641.28         91131.24     Florida
31   61136.38       152701.92         88218.23    New York
32   63408.86       129219.61         46085.25  California
33   55493.95       103057.49        214634.81     Florida
34   46426.07       157693.92        210797.67  California
35   46014.02        85047.44        205517.64    New York
36   28663.76       127056.21        201126.82     Florida
37   44069.95        51283.14        197029.42  California
38   20229.59        65947.93        185265.10    New York
39   38558.51        82982.09        174999.30  California
40   28754.33       118546.05        172795.67  California
41   27892.92        84710.77        164470.71     Florida
42   23640.93        96189.63        148001.11  California
43   15505.73       127382.30         35534.17    New York
44   22177.74       154806.14         28334.72  California
45    1000.23       124153.04          1903.93    New York
46    1315.46       115816.21        297114.46     Florida
47       0.00       135426.92             0.00  California
48     542.05        51743.15             0.00    New York
49       0.00       116983.80         45173.06  California

In [15]:
y

0     192261.83
1     191792.06
2     191050.39
3     182901.99
4     166187.94
5     156991.12
6     156122.51
7     155752.60
8     152211.77
9     149759.96
10    146121.95
11    144259.40
12    141585.52
13    134307.35
14    132602.65
15    129917.04
16    126992.93
17    125370.37
18    124266.90
19    122776.86
20    118474.03
21    111313.02
22    110352.25
23    108733.99
24    108552.04
25    107404.34
26    105733.54
27    105008.31
28    103282.38
29    101004.64
30     99937.59
31     97483.56
32     97427.84
33     96778.92
34     96712.80
35     96479.51
36     90708.19
37     89949.14
38     81229.06
39     81005.76
40     78239.91
41     77798.83
42     71498.49
43     69758.98
44     65200.33
45     64926.08
46     49490.75
47     42559.73
48     35673.41
49     14681.40
Name: Profit, dtype: float64

As the fourth column, ‘state’ contains categorical data, it should be encoded before building the model.

For enoding categorical data, if we take 3 Dummy Variable D1 for State_California, D2 for State_Florida and D3 for State_New York it will lead to a Dummy variable trap. 

Dummy Variables D1, D2 and D3 are highly correlated or multicollinear because if D1 and D2 are zero it is obvious D3 is 1. ie., if the data does not belong to State_California or State_Florida then it is sure, the data belongs to State_NewYork.

So to avoid the dummy variable trap we have to drop one dummy variable while building the model.

That can be done by X=pd.get_dummies(X, drop_first=True), which will drop one dummy variable.

In [16]:
X=pd.get_dummies(X, drop_first=True)

In [17]:
X

R&D Spend  Administration  Marketing Spend  State_Florida  State_New York
0   165349.20       136897.80        471784.10              0               1
1   162597.70       151377.59        443898.53              0               0
2   153441.51       101145.55        407934.54              1               0
3   144372.41       118671.85        383199.62              0               1
4   142107.34        91391.77        366168.42              1               0
5   131876.90        99814.71        362861.36              0               1
6   134615.46       147198.87        127716.82              0               0
7   130298.13       145530.06        323876.68              1               0
8   120542.52       148718.95        311613.29              0               1
9   123334.88       108679.17        304981.62              0               0
10  101913.08       110594.11        229160.95              1               0
11  100671.96        91790.61        249744.55              0               0
12   93863.75       127320.38        249839.44              1               0
13   91992.39       135495.07        252664.93              0               0
14  119943.24       156547.42        256512.92              1               0
15  114523.61       122616.84        261776.23              0               1
16   78013.11       121597.55        264346.06              0               0
17   94657.16       145077.58        282574.31              0               1
18   91749.16       114175.79        294919.57              1               0
19   86419.70       153514.11             0.00              0               1
20   76253.86       113867.30        298664.47              0               0
21   78389.47       153773.43        299737.29              0               1
22   73994.56       122782.75        303319.26              1               0
23   67532.53       105751.03        304768.73              1               0
24   77044.01        99281.34        140574.81              0               1
25   64664.71       139553.16        137962.62              0               0
26   75328.87       144135.98        134050.07              1               0
27   72107.60       127864.55        353183.81              0               1
28   66051.52       182645.56        118148.20              1               0
29   65605.48       153032.06        107138.38              0               1
30   61994.48       115641.28         91131.24              1               0
31   61136.38       152701.92         88218.23              0               1
32   63408.86       129219.61         46085.25              0               0
33   55493.95       103057.49        214634.81              1               0
34   46426.07       157693.92        210797.67              0               0
35   46014.02        85047.44        205517.64              0               1
36   28663.76       127056.21        201126.82              1               0
37   44069.95        51283.14        197029.42              0               0
38   20229.59        65947.93        185265.10              0               1
39   38558.51        82982.09        174999.30              0               0
40   28754.33       118546.05        172795.67              0               0
41   27892.92        84710.77        164470.71              1               0
42   23640.93        96189.63        148001.11              0               0
43   15505.73       127382.30         35534.17              0               1
44   22177.74       154806.14         28334.72              0               0
45    1000.23       124153.04          1903.93              0               1
46    1315.46       115816.21        297114.46              1               0
47       0.00       135426.92             0.00              0               0
48     542.05        51743.15             0.00              0               1
49       0.00       116983.80         45173.06              0               0

Here ‘drop_first = True’ have dropped dummy variable for State_California.

In [18]:
#Splitting testdata into X_train,X_test,y_train,y_test
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=.33,random_state=101)

In [19]:
X_train

R&D Spend  Administration  Marketing Spend  State_Florida  State_New York
30   61994.48       115641.28         91131.24              1               0
15  114523.61       122616.84        261776.23              0               1
20   76253.86       113867.30        298664.47              0               0
43   15505.73       127382.30         35534.17              0               1
38   20229.59        65947.93        185265.10              0               1
44   22177.74       154806.14         28334.72              0               0
39   38558.51        82982.09        174999.30              0               0
10  101913.08       110594.11        229160.95              1               0
49       0.00       116983.80         45173.06              0               0
25   64664.71       139553.16        137962.62              0               0
33   55493.95       103057.49        214634.81              1               0
36   28663.76       127056.21        201126.82              1               0
2   153441.51       101145.55        407934.54              1               0
27   72107.60       127864.55        353183.81              0               1
34   46426.07       157693.92        210797.67              0               0
35   46014.02        85047.44        205517.64              0               1
8   120542.52       148718.95        311613.29              0               1
19   86419.70       153514.11             0.00              0               1
29   65605.48       153032.06        107138.38              0               1
12   93863.75       127320.38        249839.44              1               0
5   131876.90        99814.71        362861.36              0               1
0   165349.20       136897.80        471784.10              0               1
28   66051.52       182645.56        118148.20              1               0
4   142107.34        91391.77        366168.42              1               0
40   28754.33       118546.05        172795.67              0               0
13   91992.39       135495.07        252664.93              0               0
9   123334.88       108679.17        304981.62              0               0
48     542.05        51743.15             0.00              0               1
23   67532.53       105751.03        304768.73              1               0
6   134615.46       147198.87        127716.82              0               0
17   94657.16       145077.58        282574.31              0               1
11  100671.96        91790.61        249744.55              0               0
31   61136.38       152701.92         88218.23              0               1

In [20]:
X_test

R&D Spend  Administration  Marketing Spend  State_Florida  State_New York
37   44069.95        51283.14        197029.42              0               0
14  119943.24       156547.42        256512.92              1               0
21   78389.47       153773.43        299737.29              0               1
32   63408.86       129219.61         46085.25              0               0
22   73994.56       122782.75        303319.26              1               0
1   162597.70       151377.59        443898.53              0               0
26   75328.87       144135.98        134050.07              1               0
46    1315.46       115816.21        297114.46              1               0
42   23640.93        96189.63        148001.11              0               0
47       0.00       135426.92             0.00              0               0
16   78013.11       121597.55        264346.06              0               0
24   77044.01        99281.34        140574.81              0               1
7   130298.13       145530.06        323876.68              1               0
45    1000.23       124153.04          1903.93              0               1
18   91749.16       114175.79        294919.57              1               0
3   144372.41       118671.85        383199.62              0               1
41   27892.92        84710.77        164470.71              1               0

In [21]:
y_train

30     99937.59
15    129917.04
20    118474.03
43     69758.98
38     81229.06
44     65200.33
39     81005.76
10    146121.95
49     14681.40
25    107404.34
33     96778.92
36     90708.19
2     191050.39
27    105008.31
34     96712.80
35     96479.51
8     152211.77
19    122776.86
29    101004.64
12    141585.52
5     156991.12
0     192261.83
28    103282.38
4     166187.94
40     78239.91
13    134307.35
9     149759.96
48     35673.41
23    108733.99
6     156122.51
17    125370.37
11    144259.40
31     97483.56
Name: Profit, dtype: float64

In [22]:
y_test

37     89949.14
14    132602.65
21    111313.02
32     97427.84
22    110352.25
1     191792.06
26    105733.54
46     49490.75
42     71498.49
47     42559.73
16    126992.93
24    108552.04
7     155752.60
45     64926.08
18    124266.90
3     182901.99
41     77798.83
Name: Profit, dtype: float64

In [23]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [24]:
#Prediciting Y from Linear regression Model
y_pred = lr.predict(X_test)

In [25]:
print(y_pred)

[ 85607.15230664 155673.62480608 118154.79468613  97580.20464701
 120001.4968002  190210.80098637 115847.98929187  61877.18925431
  68681.67859679  45768.83480162 116334.51511941 110598.34252766
 165925.6565201   46040.61874891 133649.52471753 172649.80672945
  77899.89755242]


From below difference between predicted and actual y can be seen

In [26]:
df=pd.DataFrame(data={'Predicted value':y_pred,'Actual Value':y_test})
df

Predicted value  Actual Value
37     85607.152307      89949.14
14    155673.624806     132602.65
21    118154.794686     111313.02
32     97580.204647      97427.84
22    120001.496800     110352.25
1     190210.800986     191792.06
26    115847.989292     105733.54
46     61877.189254      49490.75
42     68681.678597      71498.49
47     45768.834802      42559.73
16    116334.515119     126992.93
24    110598.342528     108552.04
7     165925.656520     155752.60
45     46040.618749      64926.08
18    133649.524718     124266.90
3     172649.806729     182901.99
41     77899.897552      77798.83

In [27]:
from sklearn import metrics
print('Mean Absolute Error      :', metrics.mean_absolute_error(y_test , y_pred))
print('Mean Squared Error       :', metrics.mean_squared_error(y_test , y_pred))
print('Root Mean Squared Error  :', np.sqrt(metrics.mean_squared_error(y_test , y_pred)))
print('R Square                 :', metrics.r2_score(y_test , y_pred))

Mean Absolute Error      : 7980.207242197569
Mean Squared Error       : 102271466.34249957
Root Mean Squared Error  : 10112.935594697494
R Square                 : 0.9387894202464991


In [28]:
print(lr.intercept_)

42891.65376236457


In [29]:
X_test.columns

Index(['R&D Spend', 'Administration', 'Marketing Spend', 'State_Florida',
       'State_New York'],
      dtype='object')

In [30]:
cdf=pd.DataFrame(lr.coef_, X_train.columns, columns=['Coefficients'])
cdf

Coefficients
R&D Spend            0.794844
Administration       0.021245
Marketing Spend      0.033484
State_Florida     5530.961448
State_New York    -347.476535

If profit is to be predicted for State_California with

R &D Spend: 132455.87
Administration: 1235674.98
Marketing Spend: 678343

Dummy Variable for both State_Florida and State_New York will be zero.
Y can be calculated as:

42891.6537623646+0.794844 * 132455.87 + 0.021245 * 1235674.98 + 0.033484 * 678343 +5530.961448 * 0 -347.476535 * 0

which is equal to 19738.96

In [31]:
lr.predict([[132455.87, 1235674.98, 678343,0,0]])

array([197138.96661191])

<a id="8"></a> <br>
# 8. Conclusion 

The dataset is used for profit prediction is for 3 US states California, Florida and New York. Before building the linear regression model, 3 US states are assigned with dummy variable. If we use all 3 dummy variable for building regression model, it will lead to multicollinearity or dummy variable trap. So in order to avoid the dummy variable trap, one dummy variable is dropped while builing regression model, which can be validated by checking number of coefficients.

<a id="9"></a> <br>
# 9. Reference

https://medium.com/nerd-for-tech/linear-regression-assumptions-eb7a760e4434

https://medium.com/nerd-for-tech/what-is-dummy-variable-trap-how-it-can-be-handled-using-python-78ec17246331